[PowerPC][ISelLowering] Support -mstack-protector-guard=tls

keith-packard opened the pull request:
Add support for using a thread-local variable with a specified offset for holding the stack guard canary value. This supports both 32- and 64- bit PowerPC targets.

This mirrors changes from #108942 but targeting PowerPC instead of RISCV. Because both of these PRs modify the same driver functions, this series is stack on top of the RISC-V one.

keith-packard committed:
[PowerPC][ISelLowering] Support -mstack-protector-guard=tls

Add support for using a thread-local variable with a specified offset
for holding the stack guard canary value.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
[clang][PowerPC] Add stack protector tests

Add tests ensuring that the driver correctly handles stack protector
guard options.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
[CodeGen][PowerPC] Add codegen test for TLS stack canary

Make sure the code generated for a TLS-relative stack guard canary
references the right location before and after intervening function
code.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
[CodeGen][PowerPC] Add codegen test for global stack canary

Make sure the code generated for a global stack guard canary
references __stack_chk_guard before and after intervening function
code.

Signed-off-by: Keith Packard <keithp@keithp.com>

nathanchance commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -3605,7 +3605,8 @@ static void RenderSSPOptions(const Driver &D, const ToolChain &TC,
     if (!EffectiveTriple.isX86() && !EffectiveTriple.isAArch64() &&
         !EffectiveTriple.isARM() && !EffectiveTriple.isThumb() &&
-        !EffectiveTriple.isRISCV())
+        !EffectiveTriple.isRISCV() && !EffectiveTriple.isPPC64() &&
+        !EffectiveTriple.isPPC32())
```
Any reason not to use `EffectiveTriple.isPPC()` instead of both `EffectiveTriple.isPPC32()` and `EffectiveTriple.isPPC64()`?

keith-packard replied:
None at all -- I started adding just PPC64 support and realized that it would be easy to just do both 32 and 64 in the same series, but didn't go back and re-factor. Thanks for catching this. I've rebased on top of main so that the RISC-V changes aren't mixed into this PR.

keith-packard committed:
[CodeGen] Provide Module to useLoadStackGuardNode

Use of LOAD_STACK_GUARD may depend upon compilation parameters; the
Module has access to all of the stack guard parameter methods.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
Ok, I think this is a lot closer now -- I'm using the non-deprecated mechanism for generating the canary load instruction, although that required reworking the guard function API a bit.

arichardson commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -5616,6 +5616,10 @@ class TargetLowering : public TargetLoweringBase {
   }
 
+protected:
+  // Simple interface for targets without a Module dependency
```
There probably aren't that many overrides here? Maybe easiest to just always override the version with the Module&?

keith-packard replied:
Yup, that makes things look cleaner. Good idea.

keith-packard commented:
btw, I'm testing this using Linux from https://git.kernel.org/pub/scm/linux/kernel/git/nathan/linux.git/log/?h=b4/powerpc-fix-stackprotector-test-clang

arichardson reviewed (approved):
This LGTM but would be good to have someone who actually knows PPC to also review it.

stewartsmith commented:
It's been a few years (and I've spent more time in 16bit x86 than powerpc for a while), but I'm pretty sure this looks good.

While the POWER ISA may be lacking vowels, the lack of `r` denoting registers in the test llvm test suite is (at least for me) doing it on hard mode :)

nathanchance commented:
With [my series to fix `arch/powerpc`](https://git.kernel.org/pub/scm/linux/kernel/git/nathan/linux.git/log/?h=b4/powerpc-fix-stackprotector-test-clang) for this implementation applied to Linux, `CONFIG_HAVE_STACKPROTECTOR` is properly set and the Linux kernel dump test module (LKDTM)’s `REPORT_STACK_CANARY` test works correctly for both 32-bit and 64-bit. I think with a PowerPC specific review, we should be ready to land this.

keith-packard commented:
> With [my series to fix `arch/powerpc`](https://git.kernel.org/pub/scm/linux/kernel/git/nathan/linux.git/log/?h=b4/powerpc-fix-stackprotector-test-clang) for this implementation applied to Linux, `CONFIG_HAVE_STACKPROTECTOR` is properly set and the Linux kernel dump test module (LKDTM)’s `REPORT_STACK_CANARY` test works correctly for both 32-bit and 64-bit. I think with a PowerPC specific review, we should be ready to land this.

Stewart reviewed the PowerPC aspects of the patch (he used to work at IBM on OPAL, firmware for OpenPOWER systems).
